src/core/components/pin-set.js

```diff
@@ -0,0 +1,271 @@
+'use strict'
+
+const multihashes = require('multihashes')
+const CID = require('cids')
+const protobuf = require('protons')
+const fnv1a = require('fnv1a')
+const dagPB = require('ipld-dag-pb')
+const DAGNode = dagPB.DAGNode
+const DAGLink = dagPB.DAGLink
+const varint = require('varint')
+const once = require('once')
+const some = require('async/some')
+
+const pbSchema = require('./pin.proto')
+
+const emptyKeyHash = 'QmdfTbBqBPQ7VNxZEYEj14VmRuZBkqFbiwReogJgS1zR1n'
+const emptyKey = multihashes.fromB58String(emptyKeyHash)
+const defaultFanout = 256
+const maxItems = 8192
+const pb = protobuf(pbSchema)
+
+function toB58String (hash) {
+  return new CID(hash).toBaseEncodedString()
+}
+
+function readHeader (rootNode) {
+  // rootNode.data should be a buffer of the format:
+  // < varint(headerLength) | header | itemData... >
+  const rootData = rootNode.data
+  const hdrLength = varint.decode(rootData)
+  const vBytes = varint.decode.bytes
+  if (vBytes <= 0) {
+    throw new Error('Invalid Set header length')
+  }
+  if (vBytes + hdrLength > rootData.length) {
+    throw new Error('Impossibly large set header length')
+  }
+  const hdrSlice = rootData.slice(vBytes, hdrLength + vBytes)
+  const header = pb.Set.decode(hdrSlice)
+  if (header.version !== 1) {
+    throw new Error(`Unsupported Set version: ${header.version}`)
+  }
+  if (header.fanout > rootNode.links.length) {
+    throw new Error('Impossibly large fanout')
+  }
+  return {
+    header: header,
+    data: rootData.slice(hdrLength + vBytes)
+  }
+}
+
+function hash (seed, key) {
+  const buf = Buffer.alloc(4)
+  buf.writeUInt32LE(seed, 0)
+  const data = Buffer.concat([
+    buf, Buffer.from(toB58String(key))
+  ])
+  return fnv1a(data.toString('binary'))
+}
+
+exports = module.exports = function (dag) {
+  const pinSet = {
+    // should this be part of `object` API?
+    hasChild: (root, childhash, callback) => {
+      let seen = {}
+      if (typeof childhash === 'object') {
+        childhash = toB58String(childhash)
+      }
+
+      function searchChildren (root, cb, _links, _checked) {
+        if (!root.links.length && _links === _checked) {
+          // all nodes have been checked
+          return cb(null, false)
+        }
+
+        some(root.links, (link, cb) => {
+          const bs58Link = toB58String(link.multihash)
+          if (bs58Link in seen) return
+          if (bs58Link === childhash) {
+            return cb(null, true)
+          }
+
+          seen[bs58Link] = true
+
+          dag.get(link.multihash, (err, res) => {
+            if (err) return cb(err)
```

Comment:
Nitpick: for consistency it would be cool to use brackets. This is not a blocker, but an easy thing to change if you do change anyway.

Reply:
Happy to; I'll look for other cases